src/main/java/nwoolcan/model/brewery/batch/BatchImpl.java

```diff
@@ -145,13 +148,38 @@ public boolean isEnded() {
         return Optional.ofNullable(this.batchEvaluation);
     }
 
+    @Override
+    public boolean isStocked() {
+        return this.stockReference != null;
+    }
+
+    private void setStockReference(final Stock stock) {
+        this.stockReference = stock;
+    }
+
+    @Override
+    public Result<Empty> stockBatchInto(final Stock stock) {
+        return Result.of(stock)
+                     .require(this::isEnded)
+                     .require(() -> !this.isStocked(), new IllegalStateException())
+                     .require(s -> s.getArticle().getUnitOfMeasure().equals(this.getCurrentSize().getUnitOfMeasure()))
+                     .peek(this::setStockReference)
+                     .toEmpty();
+    }
```

Comment:
I think you are checking a lot of thing two times when it is not needed and that the order of calls are not ok. what if some of these require do not pass and you have already created the stock and added a record there? Refer to the other comment for more details on how i would swap method calls.